cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -917,4 +931,105 @@ private boolean deleteIndexRow(Row row) {
   Scanner searchByIndex(String indexColumn, String value) {
     return indexedTable.readByIndex(Bytes.toBytes(indexColumn), Bytes.toBytes(value));
   }
+
+  /**
+   * Starts scanning/deleting of V1 table value row keys in batches. If there are no more value row keys available,
+   * it will scan history rows. If there are no value/history rows, it returns empty list.
+   *
+   * @param limit Batch size for scan/delete.
+   * @param delete Determines if operation to perform is scan or delete.
+   * @return List of {@link KeyValue} where key is timestamp for history row other wise it is null,
+   * value is {@link MetadataEntry}
+   */
+  public List<KeyValue<Long, MetadataEntry>> scanOrDeleteFromV1Table(int limit, boolean delete) {
```

Comment:
This is an awkward method name (scanOrDelete). I think it's better to have two public methods, one for scanning, one for deleting. They both can be implemented using a shared private method.

Reply:
Please see my comment further up. You probably don't need to scan in order to delete, because you already know the row keys from the first scan.